Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp

```diff
@@ -36,137 +40,67 @@
+    auto paintIntoGraphicsContext = [&](GraphicsContext& context) {
+        context.clip(initialClip);
 
         if (!contentsOpaque()) {
             context.setCompositeOperation(CompositeOperator::Copy);
-            context.fillRect(IntRect { IntPoint::zero(), tileRect.size() }, Color::transparentBlack);
+            context.fillRect(initialClip, Color::transparentBlack);
             context.setCompositeOperation(CompositeOperator::SourceOver);
         }
 
         context.translate(-tileRect.x(), -tileRect.y());
         context.scale({ contentsScale, contentsScale });
         paintGraphicsLayerContents(context, mappedTileRect);
+    };
+
+    auto paintBuffer = [&](Nicosia::Buffer& buffer) {
+        buffer.beginPainting();
+
+        GraphicsContextSkia context(sk_ref_sp(buffer.surface()));
+        paintIntoGraphicsContext(context);
 
         buffer.completePainting();
     };
 
-    static const char* enableCPURendering = getenv("WEBKIT_SKIA_ENABLE_CPU_RENDERING");
-    if (enableCPURendering && strcmp(enableCPURendering, "0")) {
-        auto buffer = Nicosia::UnacceleratedBuffer::create(tileRect.size(), contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha);
+    // Skia/GPU - accelerated rendering.
+    if (auto* acceleratedBufferPool = m_coordinator->skiaAcceleratedBufferPool()) {
+        auto* glContext = PlatformDisplay::sharedDisplayForCompositing().skiaGLContext();
+        RELEASE_ASSERT(glContext);
+        GLContext::ScopedGLContextCurrent scopedCurrent(*glContext);
+
+        auto buffer = acceleratedBufferPool->acquireBuffer(tileRect.size(), !contentsOpaque());
         paintBuffer(buffer.get());
         return buffer;
     }
 
-    auto* glContext = PlatformDisplay::sharedDisplayForCompositing().skiaGLContext();
-    RELEASE_ASSERT(glContext);
-    GLContext::ScopedGLContextCurrent scopedCurrent(*glContext);
-    auto buffer = AcceleratedBufferPool::singleton().acquireBuffer(tileRect.size(), !contentsOpaque());
+    // Skia/CPU - unaccelerated rendering.
+    auto buffer = Nicosia::UnacceleratedBuffer::create(tileRect.size(), contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha);
+
+    // Non-blocking, multi-threaded variant.
+    if (auto* unacceleratedThreadedRenderingPool = m_coordinator->skiaUnacceleratedThreadedRenderingPool()) {
+        // Threaded rendering: record display lists, and asynchronously replay them using dedicated worker threads.
+        buffer->beginPainting();
+
+        auto displayList = makeUnique<DisplayList::DisplayList>();
+        DisplayList::RecorderImpl recordingContext(*displayList, GraphicsContextState(), initialClip, AffineTransform());
+        paintIntoGraphicsContext(recordingContext);
+
+        // The painting task will notify 'buffer->completePainting()' upon async completion.
+        unacceleratedThreadedRenderingPool->postPaintingTask(initialClip, buffer, WTFMove(displayList));
```

Comment:
This comment is needed because the code is not here. If we call workerPool->postTask directly here, we have all the related code together and it's easier to follow.

Reply:
Agreed.